### PR TITLE
Make reference pattern more flexible

### PR DESCRIPTION
**Pull request by ushkarev**

### CODE REVIEW

File: mtp_transaction_uploader/patterns.py [hunk 1]
```diff
@@ -3,18 +3,30 @@
...
+        [^0-9]*                                 # skip until first digit
+        (?P<day_2>[0-9]{1,2})                   # match 1 or 2 digit day of month
+        [^0-9]*                                 # skip until next digit
+        (?P<month_2>[0-9]{1,2})                 # match 1 or 2 digit month
+        [^0-9]*                                 # skip until next digit
+        (?P<year_2>[0-9]{4}|[0-9]{2})           # match 4 or 2 digit year
+        [^A-Z]*                                 # skip until next letter
+        (?P<reference_2>[A-Z][0-9]{4}[A-Z]{2})  # match the prisoner number
+        [^A-Z]*                                 # skip until end
+    )
```
Comment (SteveMarshall) [inline mtp_transaction_uploader/patterns.py]: Can we pull the components of this out? I feel like having one regex contain the reference pattern twice is likely to end up with us inadvertently fixing bugs in one side but not the other. Maybe have `DATE_PATTERN` and `PRISONER_NUMBER_PATTERN` that you can `.format()` into `CREDIT_REF_PATTERN` to append the `_1`/`_2`? I might also be tempted to break this into two (`CREDIT_REF_DATE_SUFFIX_PATTERN` and `CREDIT_REF_DATE_PREFIX_PATTERN`) and match one or other of them? That would then make it more explicit which patterns are doing what, and why, and makes the regex a little lest nested.

File: tests/test_upload.py [hunk 2]
```diff
@@ -8,90 +8,79 @@
...
+
+    @classmethod
+    def add_unsuccessful_method(cls, name, reference):
+        def method(self):
+            self.assertEqual(upload.parse_credit_reference(reference), None)
+
+        setattr(cls, 'test_%s_does_not_parse' % name, method)
+
+
+CreditReferenceParsingTestCase.create_tests()
```
Comment (SteveMarshall) [inline tests/test_upload.py]: I'm not wild about this pattern, but I'm not sure what the alternative is right now…

Comment (ushkarev) [inline tests/test_upload.py]: it makes it easier to add additional failing and passing references and the tables of expected results are no less readable than before… i don't think there's any benefit to verbosely writing out every line as an individual method

Comment (SteveMarshall) [inline tests/test_upload.py]: Yeah, I'm happy to leave it as you have it, I just always find it somewhat clunky. As I say, not sure what better alternative there is.

File: mtp_transaction_uploader/patterns.py [hunk 3]
```diff
@@ -1,21 +1,36 @@
...
+_PRISONER_PATTERNS = {
+    'number': '''
+        (?P<number>[A-Z][0-9]{4}[A-Z]{2})  # match prisoner number
+    ''',
+    'date_of_birth': '''
+        (?P<day>[0-9]{1,2})          # match 1 or 2 digit day of month
+        [^0-9]*                      # skip until next digit
+        (?P<month>[0-9]{1,2})        # match 1 or 2 digit month
+        [^0-9]*                      # skip until next digit
+        (?P<year>[0-9]{4}|[0-9]{2})  # match 4 or 2 digit year
```
Comment (SteveMarshall) [inline mtp_transaction_uploader/patterns.py]: 😎

**Comment by SteveMarshall** [comment]
💥

**Pull request merged**